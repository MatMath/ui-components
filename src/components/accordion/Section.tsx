import * as React from 'react';
import styles from './Accordion.module.scss';
import { ArrowDownIcon } from '@atoms';
import { dashGreen00 } from '@colors';
import { getClassNames } from '@utility/cssUtils';

export interface AccordionSectionProps {
  label?: string;
  header?: React.ReactElement;
  children?: React.ReactNode;

  /** Only used by the Accordion Wrapper, not to be passed anything. */
  toggleSection?: (requestedState: boolean) => void;

  /** Whether the section is open or not, used when not wrapped in an Accordion. */
  open?: boolean;
  /** Adds an opacity to the whole section and blocks any pointer-events from passing through. */
  locked?: boolean;
  /** If the Accordion header can be toggled by the user.*/
  toggable?: boolean;

  sectionContentClassName?: string;
  lockedClassName?: string;
  sectionContainerStyle?: {
    backgroundColor?: string;
    lockedOpacity?: number;
  };
}

const Section = ({
  label,
  header,
  open,
  locked = false,
  toggable = true,
  toggleSection,
  children,
  sectionContainerStyle,
  sectionContentClassName,
  lockedClassName
}: AccordionSectionProps) => {
  const [isOpen, setIsOpen] = React.useState(open);

  React.useEffect(() => {
    setIsOpen(open);
  }, [open]);

  const onHeaderClick = () => {
    if (!toggable) return;

    if (toggleSection) {
      toggleSection(!isOpen);
    } else {
      setIsOpen(sectionIsOpen => !sectionIsOpen);
    }
  };

  let Header = header ? (
    <header onClick={onHeaderClick}>{header}</header>
  ) : (
    <header className={styles.AccordionSectionHeader} onClick={onHeaderClick}>
      <div>{label}</div>
      <div style={{ float: 'right' }}>
        <ArrowDownIcon
          height={50}
          width={30}
          iconAnimated={false}
          color={dashGreen00}
          rotate={isOpen ? 180 : 0}
        />
      </div>
    </header>
  );

  const containerStyle: { backgroundColor?: string; opacity?: number } = {};

  if (locked) {
    containerStyle.opacity = 0.5;
  }

  if (sectionContainerStyle !== undefined) {
    if (sectionContainerStyle.backgroundColor) {
      containerStyle['backgroundColor'] = sectionContainerStyle.backgroundColor;
    }
    if (sectionContainerStyle.lockedOpacity) {
      if (locked) {
        containerStyle['opacity'] = sectionContainerStyle.lockedOpacity;
      }
    }
  }

  const contentClass = getClassNames(
    styles.AccordionSectionContent,
    sectionContentClassName ? sectionContentClassName : '',
    locked
      ? lockedClassName
        ? lockedClassName
        : styles.AccordionSectionContentLocked
      : ''
  );

  return (
    <div
      className={styles.AccordionSectionContainer}
      style={{ ...containerStyle }}
    >
      {Header}
      {isOpen && (
        <div
          className={contentClass}
          style={{ pointerEvents: locked ? 'none' : 'auto' }}
        >
          <div className={styles.AccordionContentDivider} />
          {children}
        </div>
      )}
    </div>
  );
};

export { Section };
