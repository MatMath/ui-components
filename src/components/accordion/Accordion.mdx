---
name: Accordion
menu: Components
---

import { Playground, Props } from 'docz';
import { Accordion } from '@components/accordion/Accordion';
import { Section } from '@components/accordion/Section';

# Accordion

An Accordion component built by Accordion.Sections.

Notes on using the Accordion component wrapper : 
  - Wrap the Accordion.Sections and take control over its open states. 
  - When using the Accordion component use the "opened" prop on the Accordion component to specify the indexes(indexed by 0) of the Sections you want to be opened instead of using the open prop on the Section
  - If maxOneSectionOpen is passed, it overrides allSectionsOpened.

## Properties

### Accordion Properties

<Props of={Accordion} />

### Accordion.Section Properties

<Props of={Section} />

## Usage

In this example, we have Accordion.Section's wrapped in an Accordion Component. The Accordion then has fully control over the open states of its sections based on its prop configuration.

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
  }}
>
<div style={{ backgroundColor : "#F5F7F7", padding : "10px", width : "100%"}}>

  <Accordion 
    maxOneSectionOpen
    opened={[1]}
  >

    <Accordion.Section
      label="Sections can be wrapped in an Accordion"
    >
      <div style={{height : 150}}>
        The Accordion will then have say of when its sections can be opened/closed based on its configuration.
        <br/> <br/>
        For example, if you want behaviour like this Accordion where it enforces only one Section to be open at a time you can set the "maxOneSectionOpen" to true. Try opening another section!
      </div>
    </Accordion.Section>

    <Accordion.Section
      label="Have as many sections as you want!"
    >
      <div style={{height : 250, fontSize : "7px"}}>
        Up to you to define the content.
      </div>
    </Accordion.Section>

    <Accordion.Section
      label="A sections interactivitiy can be locked"
      locked
    >
      <div style={{height : 100}}>
        This content is viewable but no pointer-events will be passed to it.
      </div>
    </Accordion.Section>

    <Accordion.Section
      label="Sections can also be blocked from opening"
      locked
      toggable={false}
    >
      <div style={{height : 100}}>
        This content is not accessbile yet.
      </div>
    </Accordion.Section>

  </Accordion>

</div>
  
</Playground>

Here we have just an Accordion.Section. It is fully in control of when it is opened or closed.

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
  }}
>
<div style={{ backgroundColor : "#F5F7F7", padding : "10px", width : "100%"}}>

  <Accordion.Section
    label="Sections can exist standalone"
  >
    
    <div style={{height : 100}}>
      You don't always need to wrap Sections in an Accordion!
    </div>

  </Accordion.Section>

</div>
</Playground>

You can use the "header" prop on the Accordion.Section to add a custom Title for the section.

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
  }}
>
<div style={{ backgroundColor : "#F5F7F7", padding : "10px", width : "100%"}}>

  <Accordion.Section
    open={false}
    header={(
      <div style={{width : "100%", textAlign : "center", cursor : "pointer", height : "40px", paddingTop : "15px"}}>
        Can also define a completely custom Header as well ðŸŒ´
      </div>
    )}
  >

    <div style={{height : 100}}>
      ðŸŒ²ðŸŒ²ðŸŒ²
    </div>

  </Accordion.Section>

</div>
</Playground>
