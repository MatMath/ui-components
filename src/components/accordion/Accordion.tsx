import * as React from 'react';
import { Section } from './Section';
import styles from './Accordion.module.scss';
import { getClassNames } from '@utility/cssUtils';

interface AccordionState {
  maxOneSectionOpen?: boolean;
  allSectionsOpened?: boolean;
  oneSectionAlwaysOpen?: boolean;
  opened?: number[];
}

export interface AccordionProps extends AccordionState {
  children: React.ReactNode;
  accordionClassName?: string;
}

const Accordion = ({
  maxOneSectionOpen,
  allSectionsOpened,
  oneSectionAlwaysOpen,
  opened,
  accordionClassName,
  children
}: AccordionProps) => {
  const [openedSections, setOpenedSections] = React.useState<Set<number>>(
    new Set<number>(opened ? opened : [])
  );

  React.useEffect(() => {
    if (allSectionsOpened) {
      const allChildrenIndexes = Array.from(
        Array(children ? React.Children.toArray(children).length : 0).keys()
      );
      setOpenedSections(new Set<number>(allChildrenIndexes));
    } else if (oneSectionAlwaysOpen) {
      const openIndex = opened ? (opened.length > 0 ? opened : [0]) : [0];
      setOpenedSections(new Set<number>(openIndex));
    } else {
      setOpenedSections(new Set<number>(opened));
    }
  }, [opened, allSectionsOpened, children, oneSectionAlwaysOpen]);

  const toggleSection = (index: number, requestedOpen: boolean) => {
    if (requestedOpen) {
      if (maxOneSectionOpen) {
        setOpenedSections(new Set([index]));
      } else {
        setOpenedSections(
          prevSections => new Set<number>([...prevSections, index])
        );
      }
    } else {
      if (oneSectionAlwaysOpen) {
        if (openedSections.size !== 1) {
          setOpenedSections(prevSections => {
            prevSections.delete(index);
            return new Set<number>([...prevSections]);
          });
        }
      } else {
        setOpenedSections(prevSections => {
          prevSections.delete(index);
          return new Set<number>([...prevSections]);
        });
      }
    }
  };

  const containerClass = getClassNames(
    styles.AccordionContainer,
    accordionClassName ? accordionClassName : ''
  );

  return (
    <div className={containerClass}>
      {children
        ? React.Children.toArray(children).map((child: any, index: number) => {
            return React.cloneElement(child, {
              className: containerClass,
              open: openedSections.has(index),
              toggleSection: (requestedState: boolean) =>
                toggleSection(index, requestedState)
            });
          })
        : null}
    </div>
  );
};

Accordion.Section = Section;
export { Accordion };
