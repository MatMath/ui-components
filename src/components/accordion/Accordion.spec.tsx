import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { Accordion } from './Accordion';
import { Section } from './Section';

/**
 * Accordion Tests:
 *
 *  Opened Section would have 5 divs.
 *  Unopened Section has 3 divs.
 *
 *  These numbers are used to calcuate which sections are open vs closed.
 */

function sectionDivCount(accordion: ReactWrapper): number[] {
  return accordion.find(Section).map((sectionComponent: ReactWrapper) => {
    return sectionComponent.find('div').length;
  });
}

describe('<Accordion />', () => {
  it('should render default', () => {
    const accordionSection = mount(
      <Accordion>
        <Accordion.Section label='Some title'>Content</Accordion.Section>
      </Accordion>
    );
    expect(accordionSection).toMatchSnapshot();
  });

  it('should render a standlone section that is locked but still clickable', () => {
    const mockClick = jest.fn();
    const accordionSection = mount(
      <Accordion.Section label='Some title' locked open>
        Contet
        <button onClick={mockClick}></button>
      </Accordion.Section>
    );
    expect(
      getComputedStyle(accordionSection.getDOMNode()).getPropertyValue(
        'opacity'
      )
    ).toBe('0.5');
    expect(
      getComputedStyle(
        accordionSection
          .children()
          .childAt(1)
          .getDOMNode()
      ).getPropertyValue('pointer-events')
    ).toBe('none');
    accordionSection.simulate('click');
    accordionSection.find('button').simulate('click');
    expect(mockClick).toHaveBeenCalledTimes(1);
  });

  it('should render a standlone section that is locked and not clickable', () => {
    const accordionSection = mount(
      <Accordion.Section label='Some title' locked toggable={false}>
        Contet
      </Accordion.Section>
    );
    expect(
      getComputedStyle(accordionSection.getDOMNode()).getPropertyValue(
        'opacity'
      )
    ).toBe('0.5');
    accordionSection.simulate('click');
    expect(accordionSection.children().length).toBe(1);
  });

  it('should have them all be open', () => {
    const accordion = mount(
      <Accordion allSectionsOpened>
        <Accordion.Section label='Some title'>Content</Accordion.Section>
        <Accordion.Section label='Some title'>Content</Accordion.Section>
        <Accordion.Section label='Some title'>Content</Accordion.Section>
      </Accordion>
    );

    expect([...sectionDivCount(accordion)]).toEqual([5, 5, 5]);
  });

  it('should only have the first one opened', () => {
    const accordion = mount(
      <Accordion opened={[0]}>
        <Accordion.Section label='Some title'>Content</Accordion.Section>
        <Accordion.Section label='Some title'>Content</Accordion.Section>
        <Accordion.Section label='Some title'>Content</Accordion.Section>
      </Accordion>
    );
    expect([...sectionDivCount(accordion)]).toEqual([5, 3, 3]);
  });

  it('should be managed by the accordion parent, allowing only one section opened at a time', () => {
    const accordion = mount(
      <Accordion maxOneSectionOpen opened={[2]}>
        <Accordion.Section label='Some title' open>
          Contet
        </Accordion.Section>
        <Accordion.Section label='Some title' open>
          Contet
        </Accordion.Section>
        <Accordion.Section label='Some title'>Content</Accordion.Section>
      </Accordion>
    );
    expect([...sectionDivCount(accordion)]).toEqual([3, 3, 5]);
  });
});
