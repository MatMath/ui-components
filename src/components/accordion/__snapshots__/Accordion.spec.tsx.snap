// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> should render default 1`] = `
<Accordion>
  <div
    className="AccordionContainer"
  >
    <Section
      className="AccordionContainer"
      key=".0"
      label="Some title"
      open={false}
      toggleSection={[Function]}
    >
      <div
        className="AccordionSectionContainer"
        style={Object {}}
      >
        <header
          className="AccordionSectionHeader"
          onClick={[Function]}
        >
          <div>
            Some title
          </div>
          <div
            style={
              Object {
                "float": "right",
              }
            }
          >
            <Component
              color="#0E353D"
              height={50}
              iconAnimated={false}
              rotate={0}
              width={30}
            >
              <Component
                color="#0E353D"
                height={50}
                iconAnimated={false}
                rotate={0}
                width={30}
              >
                <svg
                  className="svgWrapper animated "
                  style={
                    Object {
                      "--icon-active-color": "#0E353D",
                      "--icon-height": "50px",
                      "--icon-hover-color": "#0E353D",
                      "--icon-idle-color": "#0E353D",
                      "--icon-rotate": "none",
                      "--icon-width": "30px",
                    }
                  }
                  version="1.1"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clipRule="evenodd"
                    d="M10 11.5503L5.63659 6L4.5 7.00399L10 14L15.5 7.00399L14.3634 6L10 11.5503Z"
                    fillRule="evenodd"
                  />
                </svg>
              </Component>
            </Component>
          </div>
        </header>
      </div>
    </Section>
  </div>
</Accordion>
`;
